prefork/prefork.go

```diff
@@ -110,30 +120,84 @@ func (p *Prefork) setTCPListenerFiles(addr string) error {
 	return nil
 }
 
-func (p *Prefork) prefork(addr string) error {
-	chErr := make(chan error, 1)
-
+func (p *Prefork) prefork(addr string) (err error) {
 	if !p.Reuseport {
-		if err := p.setTCPListenerFiles(addr); err != nil {
-			return err
+		if err = p.setTCPListenerFiles(addr); err != nil {
+			return
 		}
 
-		defer p.ln.Close()
+		defer func() {
+			if err == nil {
+				err = p.ln.Close()
+			}
+		}()
+	}
+
+	type procSig struct {
+		pid int
+		err error
 	}
 
-	for i := 0; i < runtime.GOMAXPROCS(0); i++ {
+	goMaxProcs := runtime.GOMAXPROCS(0)
+	sigCh := make(chan procSig, goMaxProcs)
+	childProcs := make(map[int]*exec.Cmd)
+
+	defer func() {
+		for _, proc := range childProcs {
+			_ = proc.Process.Kill()
+		}
+	}()
+
+	for i := 0; i < goMaxProcs; i++ {
 		/* #nosec G204 */
 		cmd := exec.Command(os.Args[0], append(os.Args[1:], preforkChildFlag)...)
 		cmd.Stdout = os.Stdout
 		cmd.Stderr = os.Stderr
 		cmd.ExtraFiles = p.files
+		if err = cmd.Start(); err != nil {
+			log.Printf("failed to start a child prefork process, error: %v\n", err)
```

Comment:
I would prefer to not use the `log` package. This was users don't have much control over the output of their program. We don't do this anywhere else in fasthttp either. Better is doing something similar to https://godoc.org/github.com/valyala/fasthttp#Server.Logger

Reply:
Got it.